core/src/main/scala/org/bitcoins/core/wallet/builder/RawTxFinalizer.scala

```diff
@@ -0,0 +1,169 @@
+                                                         changeSPK)
+    val totalCrediting = inputInfos
+      .map(_.output.value)
+      .foldLeft[CurrencyUnit](Satoshis.zero)(_ + _)
+    val totalSpending =
+      outputs.map(_.value).foldLeft[CurrencyUnit](Satoshis.zero)(_ + _)
+    val witnesses = inputInfos.map(InputInfo.getScriptWitness)
+    val txNoChangeFee = TransactionWitness.fromWitOpt(witnesses) match {
+      case _: EmptyWitness =>
+        BaseTransaction(version, inputs, outputsWithChange, lockTime)
+      case wit: TransactionWitness =>
+        WitnessTransaction(version, inputs, outputsWithChange, lockTime, wit)
+    }
+
+    val dummyTxF = Transaction.addDummySigs(txNoChangeFee, inputInfos)
+
+    val txF = dummyTxF.map { dummyTx =>
+      val fee = feeRate.calc(dummyTx)
+      val change = totalCrediting - totalSpending - fee
+      val newChangeOutput = TransactionOutput(change, changeSPK)
+      val newOutputs = if (change <= Policy.dustThreshold) {
+        outputs
+      } else {
+        outputs :+ newChangeOutput
+      }
+
+      txNoChangeFee match {
+        case btx: NonWitnessTransaction =>
+          BaseTransaction(btx.version, btx.inputs, newOutputs, btx.lockTime)
+        case WitnessTransaction(version, inputs, _, lockTime, witness) =>
+          WitnessTransaction(version, inputs, newOutputs, lockTime, witness)
+      }
+    }
+
+    txF.flatMap { tx =>
+      val passInOutChecksT =
+        NonInteractiveWithChangeFinalizer.sanityDestinationChecks(
+          expectedOutPoints = inputs.map(_.previousOutput),
+          expectedOutputs = outputs,
+          changeSPK = changeSPK,
+          finalizedTx = tx)
+
+      val passChecksT = passInOutChecksT.flatMap { _ =>
+        Transaction.sanityChecks(forSigned = false,
+                                 inputInfos = inputInfos,
+                                 expectedFeeRate = feeRate,
+                                 tx = tx)
+      }
+
+      Future.fromTry(passChecksT.map(_ => tx))
+    }
+  }
+}
+
+object NonInteractiveWithChangeFinalizer {
+
+  def sanityDestinationChecks(
+      expectedOutPoints: Vector[TransactionOutPoint],
+      expectedOutputs: Vector[TransactionOutput],
+      changeSPK: ScriptPubKey,
+      finalizedTx: Transaction): Try[Unit] = {
+    //make sure we send coins to the appropriate destinations
+    val isMissingDestination =
+      !expectedOutputs.forall(finalizedTx.outputs.contains)
+    val hasExtraOutputs =
+      if (finalizedTx.outputs.size == expectedOutputs.size) {
+        false
+      } else {
+        //the extra output should be the changeOutput
+        !(finalizedTx.outputs.size == (expectedOutputs.size + 1) &&
+          finalizedTx.outputs.map(_.scriptPubKey).contains(changeSPK))
+      }
+    val spendingTxOutPoints = finalizedTx.inputs.map(_.previousOutput)
+    val hasExtraOutPoints =
+      !spendingTxOutPoints.forall(expectedOutPoints.contains)
```

Comment:
Doesn't handle the case where `expectedOutPoints.size > spendingTxOutPoints.size`

Should just need to add a `&& expectedOutPoints.size == spendingTxOutPoints.size`